pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,282 @@
+                for j, filter in enumerate(configs):
+                    for kf, vf in filter.items():
+                        if kf == 'type':
+                            pl[f'pl_{pname}/f{j}/{kf}'] = vf
+                        else:
+                            params[kf] = vf
+
+                        if kf == 'field' and vf not in self._expr:
+                            raise KeyError(f"Missing field '{vf}' used in pipeline")
+                pl[f'pl_{pname}/f{j}/params'] = params
+
+        return pipel_cfg
+
+    def _eval_exprs(self, intg):
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        # Compute the gradients
+        if self._gradpinfo:
+            grad_soln = intg.grad_soln
+
+        # Iterate over each element type in the simulation
+        for i, (soln, ename) in enumerate(zip(intg.soln, self._enames)):
+            
+            dom_str = f'domain_{self.dom_0 + i}/'
+            # Subset and transpose the solution
+            soln = soln[...,:].swapaxes(0, 1)
+
+            mesh_op, soln_op = self.ops[ename]
+            # Interpolate the solution to the subdivided points
+            soln = soln_op @ soln
+
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln, self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns), t=intg.tcurr)
+            # subs |= dict(zip('xyz', plocs)) # TODO: add the ability for xyz in functions
+
+            # Prepare any required gradients
+            if self._gradpinfo:
+                grads = np.rollaxis(grad_soln[idx], 2)[..., :]
+
+                # Interpolate the gradients to the points
+                grad_soln = soln_op @ grad_soln
+
+                # Transform from conservative to primitive gradients
+                pgrads = self.elementscls.grad_con_to_pri(soln, grads,
+                                                          self.cfg)
+
+                # Add them to the substitutions dictionary
+                for pname, idx in self._gradpinfo:
+                    for dim, grad in zip('xyz', pgrads[idx]):
+                        subs[f'grad_{pname}_{dim}'] = grad
+
+            for field, expr in self._expr.items():
+                # Evaluate the expression at each point
+                fun = npeval(expr, subs)
+                self.mesh_n[f'{dom_str}fields/{field}/values'] = fun.T
+
+    def _get_expr(self, expr, subs={}):
+        # Ensure the expression does not contain invalid characters
+        if not re.match(r'[A-Za-z0-9_ \t\n\r.,+\-*/%()]+$', expr):
+            raise ValueError('Invalid characters in expression')
+
+        # Substitute variables
+        if subs:
+            expr = re.sub(r'\b({0})\b'.format('|'.join(subs)),
+                          lambda m: str(subs[m[1]]), expr)
+
+        # Encase in parenthesis
+        return f'({expr})'
```

Comment:
Not sure this is needed?

Reply:
which bit? Get expr is used to parse the field expressions as these are passed in a disctionary and previsiously this was only used to directly parse expression in the ini file.

Reply:
The final f string.  It just wraps the expression in () which I am unsure is needed?